src/System.Private.ServiceModel/src/System/ServiceModel/UpnEndpointIdentity.cs

```diff
@@ -53,6 +57,110 @@ internal UpnEndpointIdentity(WindowsIdentity windowsIdentity)
             _upnSid = windowsIdentity.User;
             _hasUpnSidBeenComputed = true;
         }
+
+        internal override void EnsureIdentityClaim()
+        {
+            if (_windowsIdentity != null)
+            {
+                lock (_thisLock)
+                {
+                    if (_windowsIdentity != null)
+                    {
+                        Initialize(Claim.CreateUpnClaim(GetUpnFromWindowsIdentity(_windowsIdentity)));
+                        _windowsIdentity.Dispose();
+                        _windowsIdentity = null;
+                    }
+                }
+            }
+        }
+
+        string GetUpnFromWindowsIdentity(WindowsIdentity windowsIdentity)
+        {
+            string downlevelName = null;
+            string upnName = null;
+ 
+            try
+            {
+                downlevelName = windowsIdentity.Name;
+ 
+                if (IsMachineJoinedToDomain())
+                {
+                    upnName = GetUpnFromDownlevelName(downlevelName);
+                }
+            }
+            catch (Exception e)
+            {
+                if (Fx.IsFatal(e))
+                {
+                    throw;
+                }
+             }
+ 
+            // if the AD cannot be queried for the fully qualified domain name,
+            // fall back to the downlevel UPN name
+            return upnName ?? downlevelName;
+        }
+
+        bool IsMachineJoinedToDomain()
+        {
+            throw ExceptionHelper.PlatformNotSupported();
+        }
+
+        string GetUpnFromDownlevelName(string downlevelName)
+        {
+            throw ExceptionHelper.PlatformNotSupported();
+        }
 #endif // SUPPORTS_WINDOWSIDENTITY
+
+        internal override void WriteContentsTo(XmlDictionaryWriter writer)
+        {
+            if (writer == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(writer));
+
+            writer.WriteElementString(XD.AddressingDictionary.Upn, XD.AddressingDictionary.IdentityExtensionNamespace, (string)this.IdentityClaim.Resource);
+        }
+
+#if SUPPORTS_WINDOWSIDENTITY
+        internal SecurityIdentifier GetUpnSid()
+        {
+            Fx.Assert(ClaimTypes.Upn.Equals(this.IdentityClaim.ClaimType), "");
+            if (!_hasUpnSidBeenComputed)
+            {
+                lock (_thisLock)
+                {
+                    string upn = (string)this.IdentityClaim.Resource;
+                    if (!_hasUpnSidBeenComputed)
+                    {
+                        try
+                        {
+                            NTAccount userAccount = new NTAccount(upn);
+                            _upnSid = userAccount.Translate(typeof(SecurityIdentifier)) as SecurityIdentifier;
+                        }
+#pragma warning suppress 56500 // covered by FxCOP
```

Comment:
Good point. I don't even notice those anymore. I think we should have a task to cleanup all of these littered through our code. I'll remove this instance though.